src/finch/tensor.py

```diff
@@ -38,105 +56,123 @@ def ndim(self) -> int:
         return jl.ndims(self._obj)
 
     @property
-    def shape(self) -> Tuple[int, ...]:
+    def shape(self) -> tuple[int, ...]:
         return jl.size(self._obj)
 
     @property
     def size(self) -> int:
         return np.prod(self.shape)
 
+    @property
+    def _permutation(self) -> tuple[int, ...]:
+        return jl.typeof(self._obj).parameters[1]
+
+    def _apply_new_permutation(self, value: tuple[int, ...]) -> None:
+        p = self._permutation
+        if self.order == self.ROW_MAJOR:
+            p = p[::-1]
+        new_permutation = tuple(p[i] for i in value)
+        if self.order == self.ROW_MAJOR:
+            new_permutation = new_permutation[::-1]
+        self._obj = jl.swizzle(self._obj.body, *new_permutation)
+
     def __pos__(self):
-        return Tensor(jl_data=jl.Base.broadcast(jl.seval("+"), self._obj))
+        return Tensor(jl_data=jl.Base.broadcast(jl.seval("+"), self._obj), order=self.order)
 
     def __add__(self, other):
-        return Tensor(jl_data=jl.Base.broadcast(jl.seval("+"), self._obj, other._obj))
+        return Tensor(jl_data=jl.Base.broadcast(jl.seval("+"), self._obj, other._obj), order=self.order)
 
     def __mul__(self, other):
-        return Tensor(jl_data=jl.Base.broadcast(jl.seval("*"), self._obj, other._obj))
+        return Tensor(jl_data=jl.Base.broadcast(jl.seval("*"), self._obj, other._obj), order=self.order)
 
     def __sub__(self, other):
-        return Tensor(jl_data=jl.Base.broadcast(jl.seval("-"), self._obj, other._obj))
+        return Tensor(jl_data=jl.Base.broadcast(jl.seval("-"), self._obj, other._obj), order=self.order)
 
     def __truediv__(self, other):
-        return Tensor(jl_data=jl.Base.broadcast(jl.seval("/"), self._obj, other._obj))
+        return Tensor(jl_data=jl.Base.broadcast(jl.seval("/"), self._obj, other._obj), order=self.order)
 
     def todense(self) -> np.ndarray:
-        shape = jl.size(self._obj)
+        obj = self._obj
+        shape = jl.size(obj)
 
-        # create fully dense array and access `val`
-        dense_lvls = jl.Element(jl.default(self._obj))
+        # create materialized dense array and access `val`
+        dense_lvls = jl.Element(jl.default(obj))
         for _ in shape:
             dense_lvls = jl.Dense(dense_lvls)
-        dense_tensor = jl.Tensor(dense_lvls, self._obj).lvl
+        dense_tensor = jl.Tensor(dense_lvls, obj).lvl
         for _ in shape:
             dense_tensor = dense_tensor.lvl
 
-        result = np.array(dense_tensor.val).reshape(shape, order="F")
-        return np.array(result, order="C")
+        result = np.array(jl.reshape(dense_tensor.val, shape))
+        return result.transpose(None) if self.order == self.ROW_MAJOR else result
+
+    def _is_swizzle(self) -> bool:
```

Comment:
shouldn't this always be true?

Reply:
Right now it's always true - I removed this function!